### DISCUSSION TITLE
同样的大模型，内容输出在电脑端明显比手机端快很多，可能是什么原因呢？

### DISCUSSION THREAD

**Discussion by huicewang** [lobehub/lobehub#1029] [General]
### 🥰 Feature Description

手机端访问速度明显很慢，可能原因是什么？

### 🧐 Proposed Solution

期望提升手机端内容输出速度

**Comment by arvinxx** [comment]
我感觉有可能和 tokenizer 有关系。我们会在界面上实时计算当前消息的所用 token 数，提升用户对于当前上下文窗口的感知：

<img width="444" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/7b979e45-0bf5-44a0-b680-4c71d3065401">


这个功能在实现上，会使用一个重计算的 wasm 模块 [gpt-tokenizer](https://www.npmjs.com/package/gpt-tokenizer)，用于将字符串计数成 token。目前的实现中， tokenzier 是在主进程跑的：
```ts

  const inputTokenCount = useTokenCount(input);

  const systemRoleToken = useTokenCount(systemRole);
  const chatsToken = useTokenCount(messageString);
  const toolsToken = useTokenCount(toolsString);

// useTokenCount

import { startTransition, useEffect, useState } from 'react';

import { encodeAsync } from '@/utils/tokenizer';

export const useTokenCount = (input: string = '') => {
  const [value, setNum] = useState(0);

  useEffect(() => {
    startTransition(() => {
      encodeAsync(input)
        .then(setNum)
        .catch(() => {
          // 兜底采用字符数
          setNum(input.length);
        });
    });
  }, [input]);

  return value;
};
```

虽然我们已经做了并发优化，但如果单次计算耗时仍然很久的话，还是会阻塞主进程渲染，导致UI卡顿。


可能PC端的算力够快，但部分移动端设备的计算性能并不够强，所以出现卡顿。

如果你会修改代码的话，可以试试看直接将这一行改成 `return null` ，然后再部署试试。
 https://github.com/lobehub/lobe-chat/blob/main/src/features/ChatInput/ActionBar/Token/index.tsx#L12


如果这么改之后性能有明显提升，那么说明就是这个问题导致。我们后续考虑把这部分计算移动到 service worker 中去。

**Comment by arvinxx** [comment]
经过测试发现，当 token 数量达到 10k+ 级别时，单次 token 的计算耗时会变得不可忽略。

<img width="1647" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/d22db7e5-ff1c-45bf-bfb5-f17cd00a5f90">


看来 有必要将其挪到 Web Worker 进行异步处理。

**Reply by Sun-drenched** [reply]
> 经过测试发现，当 token 数量达到 10k+ 级别时，单次 token 的计算耗时会变得不可忽略。
> 
> <img alt="image" width="1647" src="https://private-user-images.githubusercontent.com/28616219/297847551-d22db7e5-ff1c-45bf-bfb5-f17cd00a5f90.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU0MzQ5MzIsIm5iZiI6MTcyNTQzNDYzMiwicGF0aCI6Ii8yODYxNjIxOS8yOTc4NDc1NTEtZDIyZGI3ZTUtZmYxYy00NWJmLWJmYjUtZjE3Y2QwMGE1ZjkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA0VDA3MjM1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhNWY0MTJkYThjZmRhMzA3ZjI2OTZjMDIwOWRjZDc0YWM3YmM3NzcwMDMzMmIyNzQ1ZGZkN2JhMWNhMmZjNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dY02Fj1EPsB33IourMMWgFqYua4QgdfCtwMX00kJqcY">
> 看来 有必要将其挪到 Web Worker 进行异步处理。

建议给一个实时计数的flag，我觉得大多数人不在乎工具栏那跳动的数字。个人觉得可以实现成输出文字时有个转圈动画表示正在计数（输出），输出结束后尽快显示就行。

**Comment by arvinxx** [comment]
已进行第一轮优化：
- https://github.com/lobehub/lobe-chat/pull/3754
